src/state.rs

```diff
@@ -28,13 +28,14 @@ impl State {
     }
 
     fn can_move_piece(&self, piece: Piece, from: Pos, to: Pos) -> bool {
-        let to_piece = self.board.piece_at(to);
+        let (board, player) = (&self.board, self.player);
```

Comment:
is this making a deep-copy of board and player so that you don't mutate `self`?

Reply:
in this case, no. we are _borrowing_ board, it's just a reference.

`&self.board` means, borrow a reference to the value `self.board`.

a more explicit way to write this is:

```rust
let player: Player = self.player;
let board: &Board = &self.board;
```

for the `player` value by contrast, in this case we are copying it.

we still aren't able to actually mutate `board`, because a) you are correct, `&self` cannot be mutated (it would need to be `&mut self` i think), and b) you can't mutate a borrowed reference, you would need to borrow mutably, `&mut board`, which isn't possible because of a). if i try to write `let board: &mut Board = &mut self.board;`, i get the compiler error:

```
error[E0596]: cannot borrow `self.board` as mutable, as it is behind a `&` reference
  --> src/state.rs:32:33
   |
30 |     fn can_move_piece(&self, piece: Piece, from: Pos, to: Pos) -> bool {
   |                       ----- help: consider changing this to be a mutable reference: `&mut self`
31 |         let player: Player = self.player;
32 |         let board: &mut Board = &mut self.board;
   |                                 ^^^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```